Go to docs for functions

ascandone opened the pull request:
Implements https://github.com/gleam-lang/gleam/issues/2454

This PR provides a "View in hexdocs" link when hovering over public symbols from external packages
Currently only working for functions and constants (no records)

As an example, in the following file:
```gleam
import gleam/dict

pub fn f() {
  dict.delete
}
```


Hovering on `dict.delete` looks like:

<img width="522" alt="Screenshot 2023-12-04 at 14 30 42" src="https://github.com/gleam-lang/gleam/assets/26420752/18d94906-f7c1-4a7f-af55-a05806395946">

Implementation works on both qualified and unqualified imports

**current limitations:** the "view on hexdocs" link does not work on records.

lpil commented:
Hi @ascandone ! Have you linked the right issue? That one is for adding links in the generated HTML documentation.

ascandone commented:
> Hi @ascandone ! Have you linked the right issue? That one is for adding links in the generated HTML documentation.

That's true! I'll remove the issue mention from the PR description
Can the PR still be useful? It provides same functionality as elixir's "View in Hexdocs" link on hover

lpil commented:
It could be useful but we'd need to make sure to only show a link if the code exists on hexdocs. Do you have any thoughts on how the LSP would know if it is on there or not?

ascandone commented:
> It could be useful but we'd need to make sure to only show a link if the code exists on hexdocs. Do you have any thoughts on how the LSP would know if it is on there or not?

The way Elixir behaves is that this link is always shown, regardless of whether the package was published.
I don't think is doable to check if the module you are writing was published easily, however if we want to make sure there aren't any 404 links I could try to only show the link for external dependencies

lpil commented:
We definitely do not want to copy what that Elixir language server does as clicking a link and getting an error is very poor UX. We will need to only show links when the documentation is on Hex. Let's discuss how to do this in an issue and then it can be implemented once we have determined a suitable design.

ascandone commented:
> We definitely do not want to copy what that Elixir language server does as clicking a link and getting an error is very poor UX. We will need to only show links when the documentation is on Hex. Let's discuss how to do this in an issue and then it can be implemented once we have determined a suitable design.

That makes sense! I created an issue here: https://github.com/gleam-lang/gleam/issues/2454

lpil commented on compiler-core/src/language_server/engine.rs:
```diff
@@ -613,3 +640,57 @@ fn code_action_unused_imports(
+    })?;
+
+    let link = format!("https://hexdocs.pm/{package_name}/{module_name}.html#{name}");
+    Some(format!("\nView on [hexdocs]({link})"))
```
```suggestion
    Some(format!("\nView on [HexDocs]({link})"))
```

lpil commented on compiler-core/src/language_server/tests/mod.rs:
```diff
@@ -82,6 +82,21 @@ impl LanguageServerTestIO {
+            .join(dep)
+            .join("src")
+            .join(name)
+            .with_extension("gleam");
```
Please use the method on the path class rather than duplicating this logic here.

lpil commented on compiler-core/src/language_server/tests/mod.rs:
```diff
@@ -254,6 +269,44 @@ impl ProgressReporter for LanguageServerTestIO {
+        .join("build")
+        .join("packages")
+        .join(name)
+        .join("gleam.toml");
```
Please use the method on the path class rather than duplicating this logic here.

lpil commented on compiler-core/src/language_server/tests/mod.rs:
```diff
@@ -82,6 +82,21 @@ impl LanguageServerTestIO {
         path
     }
 
+    pub fn hex_dep_module(&self, dep: &str, name: &str, code: &str) -> Utf8PathBuf {
```
Could you rename `dep_module` to `path_dep_module` to make it clear the difference please.

lpil commented on compiler-core/src/language_server/tests/hover.rs:
```diff
@@ -65,6 +76,269 @@ fn(Int) -> Int
+
+    assert_eq!(
+        positioned_hover(code, Position::new(6, 3)),
+        Some(Hover {
```
It would be good to use snapshot tests with `cargo-insta` here rather than typing out the returned value, to match the majority of tests in the codebase and to reduce the amount of work maintaining the tests.

lpil commented on compiler-core/src/language_server/tests/hover.rs:
```diff
@@ -65,6 +76,269 @@ fn(Int) -> Int
 }
 
+#[test]
+fn hover_local_function() {
```
Could we get tests for different types of imports please?

For example, nested modules, aliased modules, discarded modules that would have the same name, and anything else you can think of.

I think there may be a bug in finding the right import and I would like full test coverage to verify that.

ascandone replied:
You were right! A function marked with the `@external` modifier was not behaving properly

https://github.com/gleam-lang/gleam/blob/4c436129cf3e86bd4d3aede644111582a5a2ee2a/compiler-core/src/language_server/tests/hover.rs#L223

Other than that, I tried to test as many edge cases as possible

lpil commented on compiler-core/src/language_server/engine.rs:
```diff
@@ -613,3 +640,57 @@ fn code_action_unused_imports(
         .push_to(actions);
 }
+
+fn get_expr_qualified_name(expression: &TypedExpr) -> Option<(&EcoString, &EcoString)> {
```
What is `module_name` here? Is it the name or the alias of the imported module? I'm not totally sure but I think this function mixes them up.

ascandone replied:
it should be the original name... there are tests to confirm that

ascandone commented:
Sorry about the stale PR! I've been unable to have access to the computer for few days and I'll be able to implement the suggestion later this week

lpil commented on compiler-core/src/language_server/engine.rs:
```diff
@@ -55,6 +55,8 @@ pub struct LanguageServerEngine<IO, Reporter> {
     // the usual request-response loop.
     progress_reporter: Reporter,
+
+    hex_deps: std::collections::HashSet<String>,
```
Could you document this with a comment and use a `EcoString` here please? We don't use `String` other than with library that force it.

lpil commented on compiler-core/src/language_server/engine.rs:
```diff
@@ -95,6 +97,7 @@ where
             progress_reporter,
             compiler,
             paths,
+            hex_deps: std::collections::HashSet::new(),
```
The hex deps should be set immediately here as what they are is known even before compilation.

lpil reviewed (commented):
Wonderful! Thank you very much 💜

ascandone commented:
I think I fixed everything!
I've converted the tests I wrote to snaèshot test, would you like me to convert other existing tests in the same module to snapshot tests too?

The only thing I'm not sure about is whether to set the `hex_deps` property once actually works in case you add another dependency, however I tested it locally and it looks like it does. Could there be any edge cases?

lpil reviewed (approved):
Thank you!

lpil commented:
Would you mind rebasing on main to remove the merge commit? Thank you.

If you could also update the change log that would be fab!

ascandone commented:
> Would you mind rebasing on main to remove the merge commit? Thank you.
> 
> If you could also update the change log that would be fab!

Done :+1:
Let me know if the description is clear enough

lpil reviewed (approved):
Thank you
